### DISCUSSION TITLE
Simple shorthand for iteration x times: 1..10

### DISCUSSION THREAD

**Discussion by Mashrien** [dotnet/csharplang#7306] [Language Ideas]
I apologize in advance for the mention/use of Perl here, but it's a feature there that some may be familiar with:

```
for (1 .. 10) {
    print "Num: $_ \n";
    }
```

The parameter `1 .. 10` specifies the starting and end points, whereas C#'s easiest/shortest method is:

```
foreach (int i in Enumerable.Range(1, 10)) {
    Console.WriteLine($"Num: {i}");
    }
```

Would it be possible/desirable to add such a shorthand for iterations in C#?

**Comment by FaustVX** [comment]
You can already do that with a `GetEnumerator` extension method on `Range` (or probably a nuget package already do that)
For Example: [sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMAWAbj0JMwDoAZASygEcrc8AzAezAFMAhgGMAFkQAUAN0FgiTeVCIBWOnUzoAlHgDeeIgdoBOCU00cAvnmqYAbCXREASoKgBzfgFEAHsH5QAZyYeKF19Q1QAZlp7AElPKAgAW34wQWA+AB4WYAA+IgBxfmAE5NT0vglgUSYA51cPIjT3fm1cQyIAXnzmxtqiHSIAZWBZUAGiWICAMTAeJISEIi5BABsA/hgiADU1iH4QIhk5ANGwYAstxcPBqdn56+W1ja3d1f3D46J/OCILK3aHUMAH4iKUUmkAEarfh0FwtCSnMZbH5EAC0RCR5yIiCImE0dCKJUSEIqYAkbSBhkO1TmAHciFB+AyfMJ+AAHYDBKAUyxAA)

**Comment by Mashrien** [comment]
So it's **almost** there.. Just need the enumerator extension to be built into the language itself, rather than having to add an extension for something that (in my opinion) should simply be a syntactic function/feature, not requiring additional codebehind

Edit: Though I will say I am grateful for that cute little extension, makes the code cleaner at a glance, would only be better with just a simple "0..9" without needing "var i in" before it.

**Reply by HaloFour** [reply]
The language team has expressed that the current pattern is sufficient.  Questions around clusivity were enough for both the runtime and language teams to not want to explicitly add support for this.

**Reply by Mashrien** [reply]
Can't believe the language team responded so quickly, and at such an hour.  If that's their decision then I guess you may close this suggestion / feature-request

**Reply by HaloFour** [reply]
This isn't the first time that the feature has been suggested.  Originally making ranges enumerable was a plan when that feature was being planned, but it ended up being pulled.

See:
https://github.com/dotnet/csharplang/discussions/3004
https://github.com/dotnet/runtime/issues/462

**Reply by DavidArno** [reply]
> Can't believe the language team responded so quickly, and at such an hour. If that's their decision then I guess you may close this suggestion / feature-request

Just to be clear: no one from the language team have responded. Just a bunch of enthusiastic devs that hang out here. 😃

**Reply by HaloFour** [reply]
Correct, just referring to previous discussions on the subject.  And you're right about collection literals.  I forgot about that, so thanks for bringing it up!

> Just a bunch of enthusiastic devs that hang out here. 😃

You and me both! 😄

**Reply by FaustVX** [reply]
@DavidArno @HaloFour 
> Just a bunch of enthusiastic devs that hang out here. 😃

The firsts times I read these GH issues/discussions, I was sure you worked with the LDT because of how often and well you responded (with precises infos and links)

**Reply by CyrusNajmabadi** [reply]
> So it's almost there.. Just need the enumerator extension to be built into the language itself, 

So we strongly disagree with this general sentiment. The language is the last place you want to keep loading up with kitchen sink functionality.  It's already large, complex, and delicate to move forward.  Each thing you add, increases that, which also impacts our overall velocity. 

Instead, we want to give users and library authors the tools to build composable components that then can leverage without having the language itself have to change. 

I'm this domain, for example, we already have an extensible way for anything to be iterable.   Furthermore, with "extension everything" it might even be that you could just make ranges iterable in one place and have that be picked up at all consumption sites. 

This is an example of adding powerful features that compose together, which would then enable these smaller qol areas, without the language having to special case this stuff directly.

**Reply by CyrusNajmabadi** [reply]
> If that's their decision then I guess you may close this suggestion / feature-request

It's less a decision, and more that this has been requested and our current belief is that this is better served by either an explicit extension by the user, or as something that falls out from "extension everything". Our preference is that the language not have built in special knowledge in these locations, but that more general features would allow the to "fall out".

**Comment by DavidArno** [comment]
There's a slim possibility that this functionality might happen by accident with [the upcoming collection literals feature](https://github.com/dotnet/csharplang/blob/main/proposals/collection-literals.md). The syntax might be something like:
```cs
foreach (var i in [1,..10]) 
{
    Console.WriteLine($"Num: {i}");
}
```
It's a case of "watch this space" to see how that feature pans out and whether this does become possible or not.

**Reply by CyrusNajmabadi** [reply]
Collection literals will not work in this fashion.  `10` is not iterable, so `..10` will fail to compile.

**Reply by HaloFour** [reply]
That's right, I recall reading that `foreach (var i in [1, 2, 3])` would work, but not cases involving ranges.  That's a shame, would be kind of interesting if it could be supported, but any syntax I can think of offhand already means something else....

**Reply by CyrusNajmabadi** [reply]
`foreach (var x in Range(1, 10))` would work today :-)

**Reply by DavidArno** [reply]
> Collection literals will not work in this fashion. `10` is not iterable, so `..10` will fail to compile.

That raises an interesting question: why isn't int iterable? `..10` is clearly a sequence. I get why this wasn't supported with ranges as eg `..^3` by itself is meaningless as an iteration. But collection literals would not have that issue. So maybe there's space here to not just dismiss this topic with a "*Collection literals will not work in this fashion*" comment and instead to consider making them work in this fashion... 😃

**Reply by CyrusNajmabadi** [reply]
> That raises an interesting question: why isn't int iterable?

Because you can't foreach an int. :-)

> and instead to consider making them work in this fashion... 

You could possibly make this work by using extension-everything to make ints iterable. Collections would then pick this up. But it would not be something special that collections were doing.  That's what I mean by stating that the literals themselves will not work in this fashion.

**Reply by DavidArno** [reply]
> You could possibly make this work by using extension-everything to make ints iterable. Collections would then pick this up. But it would not be something special that collections were doing. That's what I mean by stating that the literals themselves will not work in this fashion.

My thoughts on this are that `var x = [1, ..100];` seems a very compelling use case for collection literals. This could be achieved via extension-everything, or it could be achieved via a BCL update. So I'd assumed it was something that the working group would look at. The conclusion might only be "_this is something we could consider for a future release_", but it seems reasonable that it is at least considered and discussed with the runtime team.

"_Collection literals will not work in this fashion_" though sounds very much like "_we have completed ruled this out as something we want to do_", which seems odd.

**Reply by CyrusNajmabadi** [reply]
There is no intent at the language level to make ranges or collection literals have special iterable semantics. That position has not changed. If the runtime wants to change that, or if a user wants to augment with extensions is outside of our design space.  We will pick it up automatically if provided. But we do not believe we want to do anything beyond that, and no change in thinking has happened there since the last time this was discussed.

**Comment by Mashrien** [comment]
> There is no intent at the language level to make ranges or collection literals have special iterable semantics. That position has not changed. If the runtime wants to change that, or if a user wants to augment with extensions is outside of our design space. We will pick it up automatically if provided. But we do not believe we want to do anything beyond that, and no change in thinking has happened there since the last time this was discussed.

That's sad to hear, honestly. A quick shorthand method for iterating a static range (such as: `foreach ([1..51])`) would be a lovely addition to the myriad of features that makes the programmers life easier, making code itself shorter and easier to code, and almost as importantly, _read at a glance_

I sincerely hope the stance of the team will change in the future - It's a minor change for the dev team but dramatically increases readability and offers a not-insignificant QOL improvement for the coders themselves.

**Reply by CyrusNajmabadi** [reply]
> would be a lovely addition to the myriad of features that makes the programmers life easier, making code itself shorter and easier to code, and almost as importantly, read at a glance

As mentioned before, this will be picked up if teh runtime chooses to make things like ranges iterable.  Similarly, when 'extensions' ship, this would be something a user could add themselves if they so choose.  We do not think doing this directly is wise here as it is both deeply special casing very narrow patterns, and it would not compose with further work both inside and outside the language.

Note that the syntax you are referring to is very different from David's as well.  His added an value, then spread another value.  Yours adds a range.  Of the two, yours is closer to the semantics i think a user would want to convey here.  However, it would be nicest to just end up at: `foreach (var x in 1..51)`, and we should be able to get there without work within collection literals.

>  It's a minor change for the dev team 

It is absolutely *not* a minor change.  :)

>  It's a minor change for the dev team but dramatically increases readability and offers a not-insignificant QOL improvement for the coders themselves.

Note: you can already do: `foreach (var v in Range(1, 50))`.  So i think the claim of dramatic increase in readability is not really there.  I agree that its' not-insignificant.  However, this would be a lot of work for only a minor QOL improvement.  And, as mentioned before, likely something that will just 'fall out', with no specific work here, in the future.
